### PR TITLE
Encode dir obj names before expiration

### PR DESCRIPTION
**Pull request by krisis**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
Object names of directory objects qualified for ExpiredObjectAllVersions must be encoded appropriately before calling on deletePrefix on their erasure set.

e.g, a directory object and regular objects with overlapping prefixes could lead to unintended expiration of the regular objects

```
bucket/dir/ ---> directory object
bucket/dir/obj-1
```

When `bucket/dir/` qualifies for expiration, current implementation would remove regular objects under the prefix `bucket/dir/`, in this case, `bucket/dir/obj-1`.

## How to test this PR?
Setup MinIO and run the following script. At the end of the script, `mc ls -r --versions myminio/bucket/dir` must list regular objects (i.e not directory objects).
The test creates a situation where the directory object `dir` under `bucket` qualifies for expiry with overlapping regular objects.
```
#!/bin/bash

mc mb myminio/bucket
mc version enable myminio/bucket
mc anonymous set public myminio/bucket

for _ in $(seq 1 2)
do
   mc cp /etc/hosts myminio/bucket/dir/obj-1
done

curl -X PUT http://127.0.0.1:9000/bucket/dir/  -H 'X-Minio-Source-Mtime: 2024-01-13T12:02:04.115209866-08:00'

mc ilm rule add --expire-days 10 --expire-all-object-versions myminio/bucket/dir/

mc ls -r --versions myminio/bucket/dir
```

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Review by harshavardhana — COMMENTED** [review]
while we are at this can we add another `event.ObjectRemovedDelete` to be `event.ObjectRemovedDeleteAllVersions` to differentiate ExpireAll action?

**Comment by krisis** [comment]
> while we are at this can we add another `event.ObjectRemovedDelete` to be `event.ObjectRemovedDeleteAllVersions` to differentiate ExpireAll action?

It would be better if we did this in a separate PR. This involves adding more than one event name for lifecycle actions such as DeleteRestoredAction, DeleteRestoredVersionAction, etc.

**Comment by harshavardhana** [comment]
> It would be better if we did this in a separate PR. This involves adding more than one event name for lifecycle actions such as DeleteRestoredAction, DeleteRestoredVersionAction, etc.

My idea was to make this cherry-pick friendly, anyway I pushed the relevant change to this PR.

**Review by harshavardhana — APPROVED** [review]

File: cmd/batch-expire.go [hunk 1]
```diff
@@ -427,7 +427,7 @@ func batchObjsForDelete(ctx context.Context, r *BatchJobExpire, ri *batchJobInfo
 					default:
 					}
 					stopFn := globalBatchJobsMetrics.trace(batchJobMetricExpire, ri.JobID, attempts)
-					_, err := api.DeleteObject(ctx, exp.Bucket, exp.Name, ObjectOptions{
+					_, err := api.DeleteObject(ctx, exp.Bucket, encodeDirObject(exp.Name), ObjectOptions{
```
Comment (vadmeste) [inline cmd/batch-expire.go:430]: I can see this is encoded inside erasureServerPools.DeleteObject() as well, I do not think this is needed

File: cmd/data-scanner.go [hunk 2]
```diff
@@ -1240,7 +1240,7 @@ func applyExpiryOnNonTransitionedObjects(ctx context.Context, objLayer ObjectLay
 		}
 	}()
 
-	dobj, err = objLayer.DeleteObject(ctx, obj.Bucket, obj.Name, opts)
+	dobj, err = objLayer.DeleteObject(ctx, obj.Bucket, encodeDirObject(obj.Name), opts)
```
Comment (vadmeste) [inline cmd/data-scanner.go:1243]: I can see this is encoded inside erasureServerPools.DeleteObject() as well, I do not think this is needed

File: cmd/batch-expire.go (hunk 1 above)
Comment (krisis) [reply cmd/batch-expire.go:430]: In erasureServerPools.DeleteObject, encodeDirObject is called only when opts.DeletePrefix is false. In both these cases opts.DeletePrefix is true.

File: cmd/data-scanner.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/data-scanner.go:1243]: @vadmeste it is not - it is doing extra work see why this is needed from Decommission code.

File: cmd/batch-expire.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/batch-expire.go:430]: @vadmeste it is not - it is doing extra work see why this is needed from Decommission code.

**Pull request merged**